schematic_api/api/security_controller_.py

```diff
@@ -1,14 +1,48 @@
-from typing import List
+import logging
+from typing import Dict, Union
 
+from jwt import PyJWKClient, decode
+from jwt.exceptions import InvalidTokenError
+from synapseclient import Synapse
 
-def info_from_bearerAuth(token):
+from schematic.configuration.configuration import CONFIG
+
+logger = logging.getLogger(__name__)
+
+syn = Synapse(
+    configPath=CONFIG.synapse_configuration_path,
+    cache_client=False,
+)
+jwks_client = PyJWKClient(
+    uri=syn.authEndpoint + "/oauth2/jwks", headers=syn._generate_headers()
+)
+
+
+def info_from_bearerAuth(token: str) -> Dict[str, Union[str, int]]:
     """
-    Check and retrieve authentication information from custom bearer token.
-    Returned value will be passed in 'token_info' parameter of your operation function, if there is one.
-    'sub' or 'uid' will be set in 'user' parameter of your operation function, if there is one.
-    :param token Token provided by Authorization header
-    :type token: str
-    :return: Decoded token information or None if token is invalid
-    :rtype: dict | None
+    Authenticate user using bearer token. The token claims are decoded and returned.
+
+    Example from:
+    <https://pyjwt.readthedocs.io/en/stable/usage.html#retrieve-rsa-signing-keys-from-a-jwks-endpoint>
+
+    Args:
+        token (str): Bearer token.
+
+    Returns:
+        dict: Decoded token information.
     """
-    return {"uid": "user_id"}
+    try:
+        jwks_client.get_signing_key_from_jwt(token)
+        signing_key = jwks_client.get_signing_key_from_jwt(token)
+        data = decode(
+            jwt=token,
+            key=signing_key.key,
+            algorithms=["RS256"],
+            options={"verify_aud": False},
+        )
```

Comment:
The new code lacks tests.
What if the token is expired?
The returned 'data' doesn't appear to be used in any way. 
The old code returned a key 'uid' but the new code does not.
The comments say "The token claims are decoded and returned." but the code doesn't do that.
The code returns the raw token, signing key, etc. Why?
If you want to return the signing algorithm, don't hard code it, get it from the OIDC discovery document, https://repo-prod.prod.sagebase.org/auth/v1/oauth2/jwks

Reply:
> The new code lacks tests.

The code is implicitly being tested via the existing integration tests, but I agree. I will add some narrower tests for this logic specifically.

> What if the token is expired?

The pyjwt library will raise an exception:
```
    def _validate_exp(
        self,
        payload: dict[str, Any],
        now: float,
        leeway: float,
    ) -> None:
        try:
            exp = int(payload["exp"])
        except ValueError:
            raise DecodeError("Expiration Time claim (exp) must be an integer.")

        if exp <= (now - leeway):
            raise ExpiredSignatureError("Signature has expired")
```

> The returned 'data' doesn't appear to be used in any way.

The data can be returned in each API route via specifying a `token_info` argument. I am planning to make updates in a further PR to add the user ID to each open telemetry trace so we can easily identify what user produced a trace.

https://connexion.readthedocs.io/en/stable/request.html#context

> The old code returned a key 'uid' but the new code does not.

Thats correct. The old code could have returned anything (Except None) and I think the web framework would have been fine with it.

> The comments say "The token claims are decoded and returned." but the code doesn't do that.
> The code returns the raw token, signing key, etc. Why?

The code calls the `decode` function (This is Python syntax). It is returning the decoded JWT payload.
```
# The import:
from jwt import PyJWKClient, decode
.
.
.
        data = decode(
            jwt=token,
            key=signing_key.key,
            algorithms=["RS256"],
            options={"verify_aud": False},
        )
```

> If you want to return the signing algorithm, don't hard code it, get it from the OIDC discovery document, https://repo-prod.prod.sagebase.org/auth/v1/oauth2/jwks

The code is accomplishing this via this code:
```
jwks_client = PyJWKClient(
    uri=syn.authEndpoint + "/oauth2/jwks", headers=syn._generate_headers()
)
```

As a part of the JWT library it handles the logic needed to parse the document.

Reply:
My mistake, I didn't see "decode()", sorry.  While it will work to hard code `RS256` (since that's what Synapse uses), it would be best to get it from Synapse's `jwks` endpoint.

Reply:
The synapse `jwks` endpoint does not provide the `alg` so I wouldn't be able to grab it from there

Reply:
It's the `kty` (key type) field here: https://repo-prod.prod.sagebase.org/auth/v1/oauth2/jwks.
The algorithm for getting the type is:
1. get the Key ID (kid) from the token;
2. look up the kid in the list of public keys returned by the jwks endpoint.
3. for the matching record, take the kid field.

Reply:
> There may be a Python library that does all this for you, e.g. you give it the token and the jwks endpoint and it, validates the token and returns the payload (if valid).

The python library I am using is what is handling this for me, they did not have it in their examples but I was able find out that I could grab the algorithm used like:

```
        signing_key = jwks_client.get_signing_key_from_jwt(token)
        data = decode(
            jwt=token,
            key=signing_key.key,
            algorithms=[signing_key.algorithm_name],
            options={"verify_aud": False},
        )
```